helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ResourceAssignmentOptimizerAccessor.java

```diff
@@ -277,7 +296,26 @@ private AssignmentResult computeOptimalAssignmentForResources(InputFields inputF
           result);
     }
 
-    return result;
+    if (inputFields.returnFormat.equals(AssignmentFormat.CurrentStateFormat.name())) {
+      return changeToCurrentStateFormat(result);
+    } else {
+      return result;
+    }
+  }
+
+  // IdealState format   : Map of resource -> partition -> instance -> state.  (default)
+  // CurrentState format : Map of instance -> resource -> partition -> state.
+  private AssignmentResult changeToCurrentStateFormat(AssignmentResult idealStateFormatResult) {
```

Comment:
AssignmentResult is now more complicated, it may have different information. It would be best if we can return this information in the response. Considering if users are going to async call and the handling logic does not know what the original requested format is.

But either way, in the server code here, we should keep tracking what format the AssignmentResult object is using. Either adding a field (indicating format) or child classes (cleaner separate the logic but more code) is fine.

Reply:
TFTR. 
I am not sure if we need to include the return format in the response since the response itself is self explanatory. It is clear that this is a IdealState format or a CurrentState format. However, if user is doing async call with some filter, we might want to return that information.

Reply:
I don't think it is really "self-explanatory". If the controller is managing hosts as the "Helix Resource", then the resource name looks exactly the same as a host name. User will need to compare with the participant name or check with the resource name to ensure. And remember we do not require a resource must be named differently from the participant. So even the logic that I described is a guess : )

It is a very corner case, but it worths the effort to ensure our public API is crystal clear.

Reply:
As or offline discussion, it is better to add metadata (response format, user defined filter etc) in response header. Will have another PR for this since it is a separate logic.  I will add a TODO for now.